### PR TITLE
29 sprint 2 functional homepage

### PR DESCRIPTION
**Pull request by ipr17**
Sorry about the size of this pull request. I would have preferred to not make it so massive, but because of the late database integration it sort of had to happen that way. Happy to have a discussion with anyone about any of this stuff before it gets merged.

I have not integrated any of my code with the session or unit configuration code, and there is still definitely functionality that needs to be added to the home page, but I figured this is a good start for something to demonstrate to Adrian regarding the home page.

I don't think that Issue 29 (is it?) should be closed with this pull request, for the record, as there is still more functionality to add. 

Also, just to keep things clear, I have tried to comment TODO messages anywhere that I have made an assumption or some sort of simplification that needs to be changed later. (I use the TODO tree extension on VS code so that's what works for me).

Summary of changes made in this PR: 
- connected the home page to the database in general (although definitely could have made mistakes there)
- added a student_suggestions js file and a route, this makes it so that when a user types in the box to add a student, they get constantly updating student suggestions.
- made it so that when a student is added (if they haven't already been added to this session) an attendance is made for them connected to this session, and their name will appear at the top of the class list on the home page
- made it so that when a row in the class list is clicked on, it re-routes to the single student view page for that student with the correct database entry submitted

### CODE REVIEW

File: app/routes.py [hunk 1]
```diff
@@ -158,17 +190,66 @@ def save_changes():
...
         consent_status = form.consent_status.data
+        sessionID = form.session_id.data
+        
+        student = GetStudent(studentID=studentID)[0]
 
-        # Printing for debugging
-        print(f"Student Name: {student_name}")
-        print(f"Consent Status: {consent_status}")
+        if student:
+            existing_attendance = db.session.query(Attendance).filter_by(studentID=studentID, sessionID=sessionID).first() # TODO should there be a database function for this?
```
Comment (DeltaO3) [inline app/routes.py:204]: Probably should be for cleanliness imo

**Review by DeltaO3 — COMMENTED** [review]
cant fully PR this as i cant get it working on my end.
First, theres an import issue - get_perth_time is now in helpers, not utilities.
Then after self populating the database, i cant get past the following error when attempting to add student:
image
There should be better error handling for this occurence potentially, i could also be being really stupid with my database ( I hope i am). using the data from sample csv, and that unit.
Could you also potentially explain in the PR what exactly you have added/changed? I can gather it more or less from the code but am not 100% sure.

**Comment by ipr17** [comment]
Yep that GetStudent error that you are having is connected to the simplification of SessionID that I made (i.e. its 1 everywhere)

**Review by DeltaO3 — CHANGES_REQUESTED** [review]
Managed to get it working fine, looks great! a few issues tho.
- Currently, it is entirely possible to enter a student which simply does not exist, throwing an error.
- I got prompted for consent for a student that had (i believe) in the database already had consent clicked.
- I get prompted for consent when a student already exists in the database
- Currently no way to log students out? happy to leave this as out of scope of this issue its already very large, but make note of it and create a new issue
- Student page seems to not have backend functionality - again, happy to leave as out of scope of this issue, just create an issue.

I like the suggestion thing i imagine that took a lot of time to figure out, and aside from the above it works from my testing. Thanks for this!

**Comment by Ay1tsMe** [comment]
Can't access the homepage at all due to this error:
```
Traceback (most recent call last):
  File "/home/adamw/Sync/SEM22024/CITS3200 Professional Computing/Project/UWAttend/venv/lib/python3.12/site-packages/flask/app.py", line 1473, in wsgi_app
    response = self.full_dispatch_request()
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/adamw/Sync/SEM22024/CITS3200 Professional Computing/Project/UWAttend/venv/lib/python3.12/site-packages/flask/app.py", line 882, in full_dispatch_request
    rv = self.handle_user_exception(e)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/adamw/Sync/SEM22024/CITS3200 Professional Computing/Project/UWAttend/venv/lib/python3.12/site-packages/flask/app.py", line 880, in full_dispatch_request
    rv = self.dispatch_request()
         ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/adamw/Sync/SEM22024/CITS3200 Professional Computing/Project/UWAttend/venv/lib/python3.12/site-packages/flask/app.py", line 865, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/adamw/Sync/SEM22024/CITS3200 Professional Computing/Project/UWAttend/app/routes.py", line 17, in home
    current_session = GetSession(sessionID=1)[0]
                      ~~~~~~~~~~~~~~~~~~~~~~~^^^
```

**Comment by ipr17** [comment]
By the way, to get this working, make sure you have a session with sessionID=1 and all students are assigned to that session

**Comment by DeltaO3** [comment]
More details on the crash when invalid student is submitted:

**Comment by DeltaO3** [comment]
More details on invalid student crash:
image
This only happens when trying to submit a student that doesnt exist on the suggestions e.g. any keysmash. This simply should not be possible to submit.

**Review by DeltaO3 — APPROVED** [review]
LGTM!

**Pull request merged**